### PR TITLE
sql/schemachanger: reassign owned by and concurrent declarative schema changer drops can conflict

### PR DESCRIPTION
**Pull request by fqazi**
Informs: #87572

A number of bugs exist inside DROP OWNED / REASSIGNED OWNED BY inside the declarative schema changer:
1.  If a concurrent DROP operation for schema/database with a REASSIGNED OWNED BY could lead to the descriptors being dropped incorrectly by legacy schema changer jobs. This patch adds additional logic to ensure that only descriptors marked for drops are cleaned up.
2. The DROP OWNED BY support inside the declarative schema changer requires no privileges to be defined, unfortunately, this logic did not have proper error handling
3. When cleaning up back-references from schemas to the database, we skipped certain operations if the database was dropped, which is true after any cascaded DROP DATABASE statement phase. Which can lead to other issues depending on the stages executed.

### CODE REVIEW

File: pkg/sql/schema_changer.go [hunk 1]
```diff
@@ -694,7 +695,11 @@ func (sc *SchemaChanger) exec(ctx context.Context) error {
 		// Some descriptors should be deleted if they are in the DROP state.
 		switch desc.(type) {
 		case catalog.SchemaDescriptor, catalog.DatabaseDescriptor:
-			if desc.Dropped() {
+			if desc.Dropped() &&
+				((desc.DescriptorType() == catalog.Database &&
+					sc.droppedDatabaseID == desc.GetID()) ||
+					(desc.DescriptorType() == catalog.Schema &&
+						sc.droppedSchemaIDs.Contains(desc.GetID()))) {
```
Comment (postamar) [inline pkg/sql/schema_changer.go:702]: nit: consider simplifying this code by separating both cases:
```
// Some descriptors should be deleted if they are in the DROP state.
if !desc.Dropped() {
    return nil
}
switch desc.(type) {
    catalog.DatabaseDescriptor:
        if sc.droppedDatabaseID != desc.GetID() { return nil }
    case catalog.SchemaDescriptor:
        if !sc.droppedSchemaIDs.Constains(desc.GetID()) { return nil }
    default:
        return nil
}
_, err := sc.execCfg.DB.Del(ctx, catalogkeys.MakeDescMetadataKey(sc.execCfg.Codec, desc.GetID()))
return err
```

**Review by postamar — APPROVED** [review]
Nice!

**Comment by fqazi** [comment]
@postamar @TFTR!

**Pull request merged**